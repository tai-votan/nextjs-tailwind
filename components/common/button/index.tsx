import React, { FC, MouseEvent, PropsWithChildren } from "react";
import classNames from "classnames";

export interface ButtonProps {
  disabled?: boolean;
  id?: string;
  style?: React.CSSProperties;
  block?: boolean;
  danger?: boolean;
  className?: string;
  onClick?: (event: MouseEvent<HTMLButtonElement>) => void;
  type?: "primary" | "dashed" | "link" | "text" | "default";
  htmlType?: "button" | "submit" | "reset" | undefined;
}

export const Button: FC<PropsWithChildren<ButtonProps>> = (props) => {
  const { block, danger, children, className, htmlType, disabled, type, ...buttonProps } = props;

  const isPrimary = ["primary"].includes(type as string);
  const isDashed = ["dashed"].includes(type as string);
  const isDefaultOrDashed = ["default", "dashed"].includes(type as string);
  const isLink = ["link"].includes(type as string);
  const isTextOrLink = ["text", "link"].includes(type as string);

  const unlessDangerAndDisabled = !danger && !disabled;
  const IsDangerAndDisabled = danger && !disabled;

  return (
    <button
      type={htmlType}
      disabled={disabled}
      className={classNames("px-6 py-1.5 border rounded transition-all duration-300 hover:opacity-80", className, {
        "w-full": block,
        "disabled:text-slate-400 cursor-not-allowed": disabled,

        // primary button
        "text-white": isPrimary && !disabled,
        "border-orange-400 bg-orange-400": isPrimary && unlessDangerAndDisabled,
        "bg-rose-600 border-rose-600": isPrimary && IsDangerAndDisabled,

        // dashed button
        "bg-white": isDefaultOrDashed,
        "border-dashed": isDashed,
        "text-rose-600 border-rose-600": isDefaultOrDashed && IsDangerAndDisabled,
        "hover:text-orange-400 hover:border-orange-400": isDefaultOrDashed && unlessDangerAndDisabled,

        // text/link button
        "border-transparent": isTextOrLink,
        "text-sky-600": isLink && unlessDangerAndDisabled,
        "text-rose-600": isTextOrLink && IsDangerAndDisabled,
        "disabled:bg-slate-100 disabled:border-slate-200": disabled && !isTextOrLink,
      })}
      {...buttonProps}
    >
      {children}
    </button>
  );
};

Button.defaultProps = {
  disabled: false,
  block: false,
  danger: false,
  type: "default",
  htmlType: "button",
};
