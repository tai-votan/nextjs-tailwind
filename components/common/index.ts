import Typography from "./typography";

export { Input } from "./input";
export { Button } from "./button";
export { Typography };
