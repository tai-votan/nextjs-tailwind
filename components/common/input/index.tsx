import React, { ChangeEvent, HTMLInputTypeAttribute } from "react";
import classNames from "classnames";

export interface InputProps {
  disabled?: boolean;
  id?: string;
  value?: string;
  style?: React.CSSProperties;
  maxLength?: number;
  className?: string;
  placeholder?: string;
  type?: HTMLInputTypeAttribute;
  onChange?: (event: ChangeEvent<HTMLInputElement>) => void;
}

export const Input = (props: InputProps) => {
  const { type, className, ...inputProps } = props;
  return (
    <input
      type={type}
      className={classNames("border outline-0 px-2 py-1.5 rounded", className, {
        "cursor-not-allowed": props.disabled,
        "w-full": ["number", "password", "text"].includes(type as string),
        "cursor-pointer": !["number", "password", "text"].includes(type as string),
      })}
      {...inputProps}
    />
  );
};

Input.defaultProps = {
  disabled: false,
  type: "text",
};
