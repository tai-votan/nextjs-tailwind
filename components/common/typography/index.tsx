import React from "react";
import { Text } from "./text";
import { Paragraph } from "./paragraph";
import { Title } from "./title";

interface TypographyProps {
  id?: string;
  style?: React.CSSProperties;
  ["aria-label"]?: string;
  className?: string;
  italic?: boolean;
  onClick?: (e?: React.MouseEvent<HTMLDivElement>) => void;
}

export interface TextProps extends TypographyProps {
  strong?: boolean;
  onClick?: (e?: React.MouseEvent<HTMLElement>) => void;
}

export interface TitleProps extends TypographyProps {
  level?: 1 | 2 | 3 | 4 | 5;
}

export interface ParagraphProps extends TypographyProps {
  strong?: boolean;
}

export default {
  Text,
  Paragraph,
  Title,
};
