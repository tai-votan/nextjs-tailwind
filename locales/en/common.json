{
  "title": "<0>Welcome to <1></1></0>",
  "login": {
    "title": "Login",
    "username": "Username",
    "password": "Password"
  }
}
