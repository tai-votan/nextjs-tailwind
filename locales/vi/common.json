{
  "title": "<0>Chào mừng bạn đến <1></1></0>",
  "login": {
    "title": "Đăng nhập",
    "username": "Tài khoản",
    "password": "Mật khẩu"
  }
}
