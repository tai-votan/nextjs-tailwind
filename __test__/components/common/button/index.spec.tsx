import { render } from "@testing-library/react";
import { Button } from "@/components/common";

describe("Button", () => {
  it("renders button", () => {
    const { container } = render(<Button />);
    expect(container).toMatchSnapshot();
  });

  it("button is disabled", () => {
    const { container } = render(<Button disabled />);
    expect(container).toMatchSnapshot();
  });

  it("button is danger", () => {
    const { container } = render(<Button danger />);
    expect(container).toMatchSnapshot();
  });

  it("button is primary disabled", () => {
    const { container } = render(<Button type="primary" disabled />);
    expect(container).toMatchSnapshot();
  });

  it("button is primary", () => {
    const { container } = render(<Button type="primary" />);
    expect(container).toMatchSnapshot();
  });

  it("button is link", () => {
    const { container } = render(<Button type="link" />);
    expect(container).toMatchSnapshot();
  });
});
