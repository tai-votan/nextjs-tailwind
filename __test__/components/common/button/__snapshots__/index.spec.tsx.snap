// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button button is danger 1`] = `
<div>
  <button
    class="px-6 py-1.5 border rounded transition-all duration-300 hover:opacity-80 bg-white text-rose-600 border-rose-600"
    type="button"
  />
</div>
`;

exports[`Button button is disabled 1`] = `
<div>
  <button
    class="px-6 py-1.5 border rounded transition-all duration-300 hover:opacity-80 disabled:text-slate-400 cursor-not-allowed bg-white disabled:bg-slate-100 disabled:border-slate-200"
    disabled=""
    type="button"
  />
</div>
`;

exports[`Button button is link 1`] = `
<div>
  <button
    class="px-6 py-1.5 border rounded transition-all duration-300 hover:opacity-80 border-transparent text-sky-600"
    type="button"
  />
</div>
`;

exports[`Button button is primary 1`] = `
<div>
  <button
    class="px-6 py-1.5 border rounded transition-all duration-300 hover:opacity-80 text-white border-orange-400 bg-orange-400"
    type="button"
  />
</div>
`;

exports[`Button button is primary disabled 1`] = `
<div>
  <button
    class="px-6 py-1.5 border rounded transition-all duration-300 hover:opacity-80 disabled:text-slate-400 cursor-not-allowed disabled:bg-slate-100 disabled:border-slate-200"
    disabled=""
    type="button"
  />
</div>
`;

exports[`Button renders button 1`] = `
<div>
  <button
    class="px-6 py-1.5 border rounded transition-all duration-300 hover:opacity-80 bg-white hover:text-orange-400 hover:border-orange-400"
    type="button"
  />
</div>
`;
